sql/src/main/java/io/crate/operation/merge/BucketMerger.java

```diff
@@ -0,0 +1,327 @@
+        } else {
+            ArrayList<Iterator<Row>> bucketIts = getIterators(buckets);
+            emitBuckets(bucketIts);
+        }
+    }
+
+    /**
+     * multi bucket sort-merge based on iterators
+     *
+     * Page1  (first merge call)
+     *     B1      B2       B3
+     *
+     *    ->B     ->A     ->A
+     *      B       C       A
+     *                      B
+     *
+     * first iteration across all buckets:
+     *
+     *      leastRow:                   A (from b2)
+     *      equal (or also leastRow):   b3
+     *
+     *      these will be emitted
+     *
+     * second iteration:
+     *
+     *    ->B       A       A
+     *      B     ->C     ->A
+     *                      B
+     *
+     *      leastRow:   A (from b3)
+     *      equal: none
+     *
+     * third iteration:
+     *
+     *    ->B       A       A
+     *      B     ->C       A
+     *                    ->B
+     *
+     *      leastRow:   B (from b1)
+     *      equal:      B (from b2)
+     *
+     * fourth iteration:
+     *
+     *      B       A       A
+     *    ->B     ->C       A
+     *                      B
+     *
+     *      leastRow:   B (from b1)
+     *
+     * after the fourth iteration the iterator that had the leastRow (B1) will be exhausted
+     * which causes B2 (Row C) to be put into the remainingIterators which will be used if a new merge call is made
+     */
+    private void emitBuckets(ArrayList<Iterator<Row>> bucketIts) {
+        exhaustedIterators.clear();
+        bucketsWithRowEqualToLeast.clear();
+        previousRows.clear();
+
+        for (Iterator<Row> bucketIt : bucketIts) {
+            if (bucketIt.hasNext()) {
+                previousRows.add(bucketIt.next());
+            } else {
+                previousRows.add(null);
+            }
+        }
+
+        int bi = 0;
+        Row leastRow = null;
+        int leastBi = -1;
+        while (exhaustedIterators.size() < numBuckets) {
+            Row row = previousRows.get(bi);
+            if (row == null) {
+                exhaustedIterators.add(bi);
+            }
+
+            if (leastRow == null) {
+                leastRow = row;
+                leastBi = bi;
+                bi = bi < numBuckets - 1 ? bi + 1 : 0;
+            } else {
+                int compare = ordering.compare(leastRow, row);
+                if (compare < 0) {
+                    leastBi = bi;
+                    leastRow = row;
+                } else if (compare == 0) {
+                    bucketsWithRowEqualToLeast.add(bi);
+                }
+
+                bi++;
+                if (bi == numBuckets) {
+                    // looked at all buckets..
+                    boolean leastBucketItExhausted = false;
+
+                    emit(leastRow);
+
+                    // send for all other buckets that are equal to least
+                    for (IntCursor equalBucketIdx : bucketsWithRowEqualToLeast) {
+                        emit(leastRow);
```

Comment:
was an optimization because I used to copy the leastRow...  can change it but I guess it doesn't really matter because it would be an equal row anyway?